File: src/main/java/nl/jhvh/java/geometry/twodimensional/Rectangle.java
package nl.jhvh.java.geometry.twodimensional;

import java.util.Objects;

import static nl.jhvh.java.geometry.GeometryUtil.RECTANGLE_DEGREES;

public class Rectangle implements TwoDimensional {

    // Non-private and non-final for testability...
    protected Parallelogram delegateFor2Dimensional;

    private final double width;

    private final double length;

    public Rectangle(double length, double width) {
        this.delegateFor2Dimensional = new Parallelogram(length, width, RECTANGLE_DEGREES);
        this.length = length;
        this.width = width;
    }

    @Override
    public double getCircumference() {
        return delegateFor2Dimensional.getCircumference();
    }

    @Override
    public double getArea() {
        return delegateFor2Dimensional.getArea();
    }

    public double getLength() {
        return length;
    }

    public double getWidth() {
        return width;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Rectangle rectangle = (Rectangle) o;
        return delegateFor2Dimensional.equals(rectangle.delegateFor2Dimensional);
    }

    @Override
    public int hashCode() {
        return Objects.hash(delegateFor2Dimensional);
    }

    @Override
    public String toString() {
        return "Rectangle{" +
                "length=" + length +
                ", width=" + width +
                '}';
    }


    public static void main(String[] args) {
        final Rectangle rectangle = new Rectangle(10.0, 20.0);
        System.out.println("rectangle.length:        " + rectangle.getLength());
        System.out.println("rectangle.width:         " + rectangle.getWidth());
        System.out.println("rectangle.circumference: " + rectangle.getCircumference());
        System.out.println("rectangle.area:          " + rectangle.getArea());
        System.out.println();
        System.out.println("Again - rectangle.area:          " + rectangle.getArea());
        System.out.println();

        final Rectangle otherRectangle = new Rectangle(10.0, 20.0);
        boolean areEqual = rectangle.equals(otherRectangle);
        System.out.println("Are the rectangles equal? " + areEqual);
    }

}

File: src/main/kotlin/nl/jhvh/kotlin/geometry/twodimensional/Rectangle.kt
package nl.jhvh.kotlin.geometry.twodimensional

import nl.jhvh.kotlin.geometry.rectangleDegrees
import nl.jhvh.java.geometry.twodimensional.Rectangle as JavaRectangle

data class Rectangle(val length: Double, val width: Double) : TwoDimensional by Parallelogram(length, width, rectangleDegrees)

fun main() {
    val rectangle = Rectangle(10.0, 20.0)
    println("rectangle.length:        ${rectangle.length}")
    println("rectangle.width:         ${rectangle.width}")
    println("rectangle.circumference: ${rectangle.circumference}")
    println("rectangle.area:          ${rectangle.area}")
    println()
    println("Again - rectangle.width:         ${rectangle.width}")
    println()

    val otherRectangle = Rectangle(length = 10.0, width = 20.0)
    val areEqual = rectangle == otherRectangle
    println("Are the rectangles $rectangle and $otherRectangle equal? $areEqual")

    // using import alias: import nl.jhvh.java.geometry.twodimensional.Rectangle as JavaRectangle
    val javaRectangle = JavaRectangle(10.0, 20.0)
    // Import aliases are nice if you have to use, say, joda LocalDateTime and java LocalDateTime in the same class
    // (or even the old fashioned java.util.Date and java.sql.Date)

    println()
    println()
    var areKotlinAndJavaRectangleEqual = rectangle.equals(javaRectangle)
    println("Are the rectangles $rectangle (Kotlin) and $javaRectangle (Java) equal? $areKotlinAndJavaRectangleEqual")

    areKotlinAndJavaRectangleEqual = javaRectangle.equals(rectangle)
    println("Are the rectangles $javaRectangle (Java) and $rectangle (Kotlin) equal? $areKotlinAndJavaRectangleEqual")
}
